surface/include/pcl/surface/3rdparty/poisson4/poisson_exceptions.h

```diff
@@ -0,0 +1,151 @@
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#pragma once 
+
+#include <stdexcept>
+#include <sstream>
+#include <boost/current_function.hpp>
+
+/** POISSON_THROW_EXCEPTION is a helper macro to be used for throwing exceptions. e.g.
+  * POISSON_THROW_EXCEPTION (PoissonBadArgumentException, "[ERROR] B-spline up-sampling not supported for degree " << Degree);
+  * 
+  * \note 
+  * Adapted from PCL_THROW_EXCEPTION. We intentionally do not reuse PCL_THROW_EXCEPTION here
+  * to avoid introducing any dependencies on PCL in this 3rd party module.       
+  */
+#define POISSON_THROW_EXCEPTION(ExceptionName, message)                     \
+{                                                                           \
+  std::ostringstream s;                                                     \
+  s << message;                                                             \
+  throw ExceptionName(s.str(), __FILE__, BOOST_CURRENT_FUNCTION, __LINE__); \
+}
+
+namespace pcl
+{
+  namespace poisson
+  {
+    /** \class PoissonException
+      * \brief A base class for all poisson exceptions which inherits from std::runtime_error
+      * 
+      * \note 
+      * Adapted from PCLException. We intentionally do not reuse PCLException here
+      * to avoid introducing any dependencies on PCL in this 3rd party module.       
+      */
+    class PoissonException : public std::runtime_error
+    {
+      public:
+        PoissonException (const std::string& error_description,
+                      const char* file_name = NULL,
+                      const char* function_name = NULL,
+                      unsigned line_number = 0)
+          : std::runtime_error (createDetailedMessage (error_description,
+                                                       file_name,
+                                                       function_name,
+                                                       line_number))
+          , file_name_ (file_name)
+          , function_name_ (function_name)
+          , line_number_ (line_number)
+        {}
+
+      protected:
+        static std::string
+        createDetailedMessage (const std::string& error_description,
+                               const char* file_name,
+                               const char* function_name,
+                               unsigned line_number)
+        {
+          std::ostringstream sstream;
+          if (function_name != NULL)
+            sstream << function_name << " ";
```

Comment:
Another clang tidy remark. `" "` is a single character, so replace it with `' '`.